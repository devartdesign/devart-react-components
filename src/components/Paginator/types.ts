export enum LinkType {
  BUTTON,
  LINK
}

export interface IPaginatorQuery {
  page: number;
  limit: number;
}
