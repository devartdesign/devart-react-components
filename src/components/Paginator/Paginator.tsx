import React, { memo, useCallback, useState } from 'react';

import Pagination from './Pagination';
// import PaginationToolkit from './Toolkit';

import { LinkType, IPaginatorProps, IPaginatorState } from './types';
import './Paginator.scss';

const Paginator = ({
  totalRecords,
  pageLimit = 1,
  pageNeighbours = 2,
  linkType = LinkType.BUTTON,
  filterOptions = [],
  hideColumnOptions = [],
  enableToolkit = false,
  hidePaginationLimiter = false,
  disableFilter = false,
  disableHideColumns = false,
  disableSearch = false,
  handleChange,
  children
}: React.PropsWithChildren<IPaginatorProps>) => {
  const [query, setQuery] = useState<IPaginatorState>({});
  const onChange = useCallback((newQuery) => {
    setQuery(prevQuery => ({ ...prevQuery, ...newQuery }));
    handleChange(newQuery);
  }, [setQuery, handleChange]);
  return (
    <div className="paginator">
      <Pagination
        totalRecords={totalRecords}
        pageLimit={pageLimit}
        pageNeighbours={pageNeighbours}
        onPageChanged={onChange}
        query={query}
        hidePaginationLimiter={hidePaginationLimiter}
        linkType={linkType}
        render={paginationProps => (
          <>
            {enableToolkit && (
              <div>Toolkit</div>
              // <PaginationToolkit
              //   onChange={onChange}
              //   filterOptions={filterOptions}
              //   hideColumnOptions={hideColumnOptions}
              //   paginationProps={paginationProps}
              //   disableFilter={disableFilter}
              //   disableHideColumns={disableHideColumns}
              //   disableSearch={disableSearch}
              // />
            )}
            {children}
          </>
        )}
      />
    </div>
  );
};

export default memo(Paginator);
