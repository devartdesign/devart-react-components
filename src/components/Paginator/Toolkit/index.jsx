export { default } from './Toolkit.container';
