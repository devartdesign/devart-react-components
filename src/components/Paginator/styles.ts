import { StyleSheet } from 'aphrodite/no-important';

export default StyleSheet.create({
  paginationNav: {
    display: 'flex',
    flexDirection: 'row',
    flexBasis: '100%',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  pageLimiter: {
    display: 'flex',
    width: '65px',
    height: '40px',
    backgroundColor: '#fff',
    borderRadius: '6px',
    boxShadow: '0 0 5px rgba(0, 0, 0, 0.1)',
  },
  navPagination: {
    display: 'flex',
    flexBasis: '100%;',
  },
  pagination: {
    backgroundColor: '#FFF',
    padding: '0 10px',
    margin: '10px 0',
    display: 'flex',
    flexGrow: 1,
    listStyleType: 'none',
  },
  pageItem: {
    display: 'flex',
    justifyContent: 'center',
    alignContent: 'center',
    width: 40,
    height: 25,
  },
  pageItemButton: {
    cursor: 'pointer',
    border: 'none',
    backgroundColor: '#FFF',
    width: '100%',
    height: '100%',
    padding: '0',
    ':hover': {
      backgroundColor: '#546e7a',
      color: '#FFF',
    }
  },
  pageItemActive: {
    backgroundColor: '#29434e',
    color: '#FFF',
  }
});
