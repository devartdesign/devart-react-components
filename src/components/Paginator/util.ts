export interface IPaginationRange {
  from: number;
  to: number;
  step?: number;
}

export const createPaginationRange = ({ from, to, step = 1 }: IPaginationRange) => {
  let i = from;
  const rangeNumber = [];
  while (i <= to) {
    rangeNumber.push(i);
    i += step;
  }
  return rangeNumber;
};
