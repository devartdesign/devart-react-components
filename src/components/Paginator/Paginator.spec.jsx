import React from 'react';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';

import { getInitialState } from '../../../mocks/store';
import Paginator from './Paginator';

describe('<Paginator />', () => {
  let wrapper;
  const defaultProps = {
    handleChange: jest.fn(),
    children: (<div>Some paginated children</div>),
    enableToolkit: false,
    totalRecords: 10,
    pageLimit: 5,
    pageNeighbours: 2,
    filterOptions: []
  };

  describe('render', () => {
    beforeEach(() => {
      wrapper = shallow(<Paginator {...defaultProps} />);
    });

    it('should render a <Pagination> with 2 pages with currentPage 1 and children over it', () => {
      expect(renderer.create(wrapper)
        .toJSON())
        .toMatchSnapshot();
    });
  });

  describe('toolkit enable', () => {
    beforeEach(() => {
      defaultProps.enableToolkit = true;
      defaultProps.filterOptions = [{
        id: 'name',
        value: 'name',
        text: 'name'
      }, {
        id: 'lastname',
        value: 'lastname',
        text: 'lastname'
      }];
      const mockStore = configureStore();
      const store = mockStore(getInitialState());
      wrapper = mount(
        <Provider store={store}>
          <Paginator {...defaultProps} />
        </Provider>
      );
    });

    it('should render <Pagination> with children above and <Search> and <Filter> toolkit components', () => {
      expect(renderer.create(wrapper)
        .toJSON())
        .toMatchSnapshot();
    });
  });

  describe('handleChange', () => {
    beforeEach(() => {
      wrapper = shallow(<Paginator {...defaultProps} />);
    });

    it('should call handleChange callback from Props', () => {
      wrapper.instance().handleChange({ currentPage: 1, pageLimit: 2 });
      expect(defaultProps.handleChange).toHaveBeenCalledWith({ currentPage: 1, pageLimit: 2 });
    });
  });
});
