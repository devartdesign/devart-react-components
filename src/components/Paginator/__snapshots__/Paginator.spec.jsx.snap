// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paginator /> render should render a <Pagination> with 2 pages with currentPage 1 and children over it 1`] = `
<div
  className="paginator"
>
  <div>
    Some paginated children
  </div>
  <div
    className="pagination__nav"
  >
    <div
      className="MuiInputBase-root-21 MuiOutlinedInput-root-8 pageLimiter"
      onClick={[Function]}
    >
      <fieldset
        aria-hidden={true}
        className="MuiPrivateNotchedOutline-root-38 MuiOutlinedInput-notchedOutline-15"
        style={
          Object {
            "paddingLeft": 8,
          }
        }
      >
        <legend
          className="MuiPrivateNotchedOutline-legend-39"
          style={
            Object {
              "width": 0.01,
            }
          }
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </legend>
      </fieldset>
      <div
        className="MuiSelect-root-1"
      >
        <select
          className="MuiSelect-select-2 MuiInputBase-input-31 MuiOutlinedInput-input-16"
          id="outlined-age-native-simple"
          multiple={false}
          name="Limit"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          value={5}
        >
          <option
            value={5}
          >
            5
          </option>
          <option
            value={10}
          >
            10
          </option>
          <option
            value={15}
          >
            15
          </option>
          <option
            value={20}
          >
            20
          </option>
          <option
            value={25}
          >
            25
          </option>
        </select>
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-40 MuiSelect-icon-7"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Paginator /> toolkit enable should render <Pagination> with children above and <Search> and <Filter> toolkit components 1`] = `
<div
  className="paginator"
>
  <div
    className="paginator__toolkit"
  >
    <div
      className="paginator__toolkit__item"
    >
      <div
        className="Search-search-49"
      >
        <div
          className="Search-searchIcon-50"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-40"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
        <div
          className="MuiInputBase-root-21 MuiInput-root-53 Search-inputRoot-51 MuiInput-underline-57"
          onClick={[Function]}
        >
          <input
            className="MuiInputBase-input-31 MuiInput-input-61 Search-inputInput-52"
            name="search"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Searchâ€¦"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="paginator__toolkit__item"
    >
      <div
        className="MuiFormControl-root-66"
        style={
          Object {
            "width": 100,
          }
        }
      >
        <label
          className="MuiFormLabel-root-81 MuiInputLabel-root-70 MuiInputLabel-formControl-75 MuiInputLabel-animated-78"
          data-shrink={false}
          htmlFor="select-multiple-checkbox"
          style={
            Object {
              "left": 4,
              "top": -10,
            }
          }
        >
          Filters
        </label>
        <div
          className="MuiInputBase-root-21 MuiInput-root-53 MuiInput-underline-57 MuiInputBase-formControl-22 MuiInput-formControl-54"
          onClick={[Function]}
          style={
            Object {
              "margin": 5,
            }
          }
        >
          <div
            className="MuiSelect-root-1"
          >
            <div
              aria-haspopup="true"
              aria-pressed="false"
              className="MuiSelect-select-2 MuiSelect-selectMenu-5 MuiInputBase-input-31 MuiInput-input-61"
              id="select-filters"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </div>
            <input
              id="select-multiple-checkbox"
              name="filters"
              placeholder="Filters"
              type="hidden"
              value=""
            />
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-40 MuiSelect-icon-7"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      className="paginator__toolkit__item"
    >
      <div
        className="MuiFormControl-root-66"
        style={
          Object {
            "width": 140,
          }
        }
      >
        <label
          className="MuiFormLabel-root-81 MuiInputLabel-root-70 MuiInputLabel-formControl-75 MuiInputLabel-animated-78"
          data-shrink={false}
          htmlFor="select-multiple-checkbox"
          style={
            Object {
              "left": 4,
              "top": -10,
            }
          }
        >
          Hide Columns
        </label>
        <div
          className="MuiInputBase-root-21 MuiInput-root-53 MuiInput-underline-57 MuiInputBase-formControl-22 MuiInput-formControl-54"
          onClick={[Function]}
          style={
            Object {
              "margin": 5,
            }
          }
        >
          <div
            className="MuiSelect-root-1"
          >
            <div
              aria-haspopup="true"
              aria-pressed="false"
              className="MuiSelect-select-2 MuiSelect-selectMenu-5 MuiInputBase-input-31 MuiInput-input-61"
              id="select-hideColumns"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </div>
            <input
              id="select-multiple-checkbox"
              name="hideColumns"
              placeholder="Filters"
              type="hidden"
              value=""
            />
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-40 MuiSelect-icon-7"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    Some paginated children
  </div>
  <div
    className="pagination__nav"
  >
    <div
      className="MuiInputBase-root-21 MuiOutlinedInput-root-8 pageLimiter"
      onClick={[Function]}
    >
      <fieldset
        aria-hidden={true}
        className="MuiPrivateNotchedOutline-root-38 MuiOutlinedInput-notchedOutline-15"
        style={
          Object {
            "paddingLeft": 8,
          }
        }
      >
        <legend
          className="MuiPrivateNotchedOutline-legend-39"
          style={
            Object {
              "width": 0.01,
            }
          }
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </legend>
      </fieldset>
      <div
        className="MuiSelect-root-1"
      >
        <select
          className="MuiSelect-select-2 MuiInputBase-input-31 MuiOutlinedInput-input-16"
          id="outlined-age-native-simple"
          multiple={false}
          name="Limit"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          value={5}
        >
          <option
            value={5}
          >
            5
          </option>
          <option
            value={10}
          >
            10
          </option>
          <option
            value={15}
          >
            15
          </option>
          <option
            value={20}
          >
            20
          </option>
          <option
            value={25}
          >
            25
          </option>
        </select>
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-40 MuiSelect-icon-7"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
